libraries/AP_Winch/AP_Winch_Daiwa.cpp

```diff
@@ -0,0 +1,227 @@
+{
+    AP::logger().Write_Winch(
+            healthy(),
+            latest.thread_end,
+            latest.moving,
+            latest.clutch,
+            (uint8_t)config.control_mode,
+            config.length_desired,
+            get_current_length(),
+            config.rate_desired,
+            latest.tension_corrected,
+            latest.voltage,
+            constrain_int16(latest.motor_temp, INT8_MIN, INT8_MAX));
+}
+
+// read incoming data from winch and update intermediate and latest structures
+void AP_Winch_Daiwa::read_data_from_winch()
+{
+    // return immediately if serial port is not configured
+    if (uart == nullptr) {
+        return;
+    }
+
+    // read any available characters from serial port and send to GCS
+    int16_t nbytes = uart->available();
+    while (nbytes-- > 0) {
+        int16_t b = uart->read();
+
+        if ((b >= '0' && b <= '9') || (b >= 'A' && b <= 'F') || (b >= 'a' && b <= 'f')) {
+            // add digits to buffer
+            buff[buff_len] = b;
+            buff_len++;
+            if (buff_len >= buff_len_max) {
+                buff_len = 0;
+                parse_state = ParseState::WAITING_FOR_TIME;
+            }
+        } else if (b == ',' || b == '\r') {
+            // comma or carriage return signals end of current value
+            buff[buff_len] = '\0';
+            long int ret = (int32_t)strtol(buff, nullptr, 16);
+            if (ret >= (long)INT32_MAX || ret <= (long)INT32_MIN) {
+                // failed to get valid number, throw away packet
+                parse_state = ParseState::WAITING_FOR_TIME;
+            } else {
+                // parse number received and empty buffer
+                buff_len = 0;
+                switch (parse_state) {
+                case ParseState::WAITING_FOR_TIME:
+                    intermediate.time_ms = (uint32_t)ret;
+                    parse_state = ParseState::WAITING_FOR_SPOOL;
+                    break;
+                case ParseState::WAITING_FOR_SPOOL:
+                    intermediate.line_length = (int32_t)ret * line_length_correction_factor;
+                    parse_state = ParseState::WAITING_FOR_TENSION1;
+                    break;
+                case ParseState::WAITING_FOR_TENSION1:
+                    intermediate.tension_uncorrected = (uint16_t)ret;
+                    parse_state = ParseState::WAITING_FOR_TENSION2;
+                    break;
+                case ParseState::WAITING_FOR_TENSION2:
+                    intermediate.tension_corrected = (uint16_t)ret;
+                    parse_state = ParseState::WAITING_FOR_THREAD_END;
+                    break;
+                case ParseState::WAITING_FOR_THREAD_END:
+                    intermediate.thread_end = (ret > 0);
+                    parse_state = ParseState::WAITING_FOR_MOVING;
+                    break;
+                case ParseState::WAITING_FOR_MOVING:
+                    intermediate.moving = constrain_int16(ret, 0, UINT8_MAX);
```

Comment:
Pretty sure `constrain_int16` should be `constrain_uint32` - i.e. the function name should match the input type, which is (currently) 32-bits long.

If ret here was `INT16_MAX+1` then moving would get a value of 0 here rather than 255 as the thing gets cast to int16_t as part of the function call, wraps into a negative number and then gets constrained.

Reply:
yes, switched to constrain_int32, txs.